main.go

```diff
@@ -196,19 +199,67 @@ func StartPurgeLoop(secInterval int) {
 	}
 }
 
-func writeToPumps(keys []interface{}, job *health.Job, startTime time.Time) {
+func writeToPumps(keys []interface{}, job *health.Job, startTime time.Time, purgeDelay int) {
 	// Send to pumps
 	if Pumps != nil {
+		var wg sync.WaitGroup
+		wg.Add(len(Pumps))
+
 		for _, pmp := range Pumps {
-			log.WithFields(logrus.Fields{
-				"prefix": mainPrefix,
-			}).Debug("Writing to: ", pmp.GetName())
-			pmp.WriteData(keys)
-			if job != nil {
-				job.Timing("purge_time_"+pmp.GetName(), time.Since(startTime).Nanoseconds())
-			}
+			go func(wg *sync.WaitGroup, pmp pumps.Pump, keys *[]interface{}) {
```

Comment:
This looks very intimidating because the closure body spans more than 3/4 lines. Lets make this into a separate function so the loop can look like this , I use `execPump` as example function name.

```go
		for _, pmp := range Pumps {
			go execPump(&wg, purgeDelay, startTime, job, pmp, &keys)
		}
```

Then `execPump` will be

```go
func execPump(wg *sync.WaitGroup, purgeDelay int, startTime time.Time, job *health.Job, pmp pumps.Pump, keys *[]interface{}) {
	timer := time.AfterFunc(time.Duration(purgeDelay)*time.Second, func() {
		if pmp.GetTimeout() == 0 {
			log.WithFields(logrus.Fields{
				"prefix": mainPrefix,
			}).Warning("Pump  ", pmp.GetName(), " is taking more time than the value configured of purge_delay. You should try to set a timeout for this pump.")
		} else if pmp.GetTimeout() > purgeDelay {
			log.WithFields(logrus.Fields{
				"prefix": mainPrefix,
			}).Warning("Pump  ", pmp.GetName(), " is taking more time than the value configured of purge_delay. You should try lowering the timeout configured for this pump.")
		}
	})
	defer timer.Stop()
	defer wg.Done()
	ch := make(chan error, 1)

	//Load pump timeout
	timeout := pmp.GetTimeout()
	var ctx context.Context
	var cancel context.CancelFunc
	//Initialize context depending if the pump has a configured timeout
	if timeout > 0 {
		ctx, cancel = context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
	} else {
		ctx, cancel = context.WithCancel(context.Background())
	}

	defer cancel()

	go func(ch chan error, ctx context.Context, pmp pumps.Pump, keys *[]interface{}) {
		ch <- pmp.WriteData(ctx, *keys)
	}(ch, ctx, pmp, keys)

	select {
	case err := <-ch:
		if err != nil {
			log.WithFields(logrus.Fields{
				"prefix": mainPrefix,
			}).Warning("Error Writing to: ", pmp.GetName(), " - Error:", err)
		}
	case <-ctx.Done():
		switch ctx.Err() {
		case context.Canceled:
			// DO something?
		case context.DeadlineExceeded:
			log.WithFields(logrus.Fields{
				"prefix": mainPrefix,
			}).Warning("Timeout Writing to: ", pmp.GetName())
		}

	}

	if job != nil {
		job.Timing("purge_time_"+pmp.GetName(), time.Since(startTime).Nanoseconds())
	}

}
```

Reply:
Solved